### PR TITLE
fix LR finder with AMP

### LINKED ISSUE #2804: AttributeError on lr_find with half precision
## 🐛 lr_find crash with mixed precision

initial run of lr_find raise AttributeError when run with 16Bit precison

### To Reproduce

1. Init module and trainer with precision=16
2. Run trainer.lr_find(module)

`AttributeError: 'NoneType' object has no attribute 'state_dict'`

Stacktrace:

```
/usr/local/lib/python3.6/dist-packages/pytorch_lightning/trainer/training_io.py in dump_checkpoint(self, weights_only)
    360             # save native amp scaling
    361             if self.use_amp and NATIVE_AMP_AVALAIBLE and not self.use_tpu:
--> 362                 checkpoint['native_amp_scaling_state'] = self.scaler.state_dict()
    363 
```

#### Code sample
```
module = new LightningModule(...)
trainer = Trainer(max_epochs=10, precision=16)
trainer.lr_find(module, min_lr=1e-06, max_lr=1e-01, early_stop_threshold=None)
```

### Environment

You can get the script and run it with Torch 1.6.0+cu101, Torchvision 0.7.0+cu101, CUDA 10.1, Lightning 0.8.5

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

I was no able to replicate the original issue but with an added test, I found another which is fixed...
Closes #2804

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@SimonNarendorf mind check that this is the case you describe in #2804

**Comment by SkafteNicki** [comment]
@Borda did you fix it?

**Comment by Borda** [comment]
> @Borda did you fix it?

Well, I was not able to find the original error (probably due to refactoring) but eventually, with adding these new test I found another similar which I have fixed...

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -277,11 +280,11 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
             if self.trainer.amp_backend == AMPType.NATIVE and not self.trainer.use_tpu and self.trainer.scaler is not None:
                 checkpoint['native_amp_scaling_state'] = self.trainer.scaler.state_dict()
             elif self.trainer.amp_backend == AMPType.APEX:
+                # not sure if we should init it or rather skip it it was not used...
+                amp.initialize(model.cuda().float())
```
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:284]: true, I was checking if there is a method that would return if it is initialized or not...
we may consider it as hotfix and then add a note/ that we need a better check for this case
other way would be to wrap it in try/catch and init only if needed...

[the remark this answers is not in the document]
Comment (SeanNaren) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:284]: Agreed, I think on average this is fine

[the remark this answers is not in the document]
Comment (SkafteNicki) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:284]: This is purely a fix such that the initial model can be saved before lr finder actually starts (which again will call `amp.initialize`). So in this case it does not really matter what `amp.state_dict()` is, as long as it exist.

**Review by SeanNaren — APPROVED** [review]

**Review by SkafteNicki — COMMENTED** [review]
LGTM

**Review by SkafteNicki — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:284]: I just think too add to the connector or model extra state if AMP was initialized and inf not just skip that state saving, does it make sense? @SkafteNicki

[the remark this answers is not in the document]
Comment (SkafteNicki) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:284]: I like that solution :]

**Comment by SkafteNicki** [comment]
@Borda why was this PR closed? It seemed to solve the incompatibility between lr finder and apex

**Comment by Borda** [comment]
I know but @williamFalcon concluded that it does not cover the root of the problem and rather solves it on his own :]

**Pull request closed without merging**